client/src/main/java/org/apache/falcon/client/FalconClient.java

```diff
@@ -1070,66 +1058,90 @@ public String getExtensionDescription(final String extensionName)  {
+                MediaType.APPLICATION_OCTET_STREAM_TYPE));
+        try {
+            formDataMultiPart.close();
+        } catch (IOException e) {
+            OUT.get().print("Submit failed. Failed to submit entities");
+            throw new FalconCLIException("Submit failed. Failed to submit entities", e);
+        }
+        return formDataMultiPart;
+    }
+
+    private List<Entity> validateExtensionAndGetEntities(String extensionName, String jobName, InputStream configStream) {
         ClientResponse clientResponse = new ResourceBuilder()
-                .path(ExtensionOperations.DETAIL.path)
+                .path(ExtensionOperations.DETAIL.path, extensionName)
                 .call(ExtensionOperations.DETAIL);
-        JSONObject responseJson = clientResponse.getEntity(JSONObject.class);
-        ExtensionType extensionType;
+        JSONObject responseJson;
+        try {
+            responseJson = new JSONObject(clientResponse.getEntity(String.class));
+        } catch (JSONException e) {
+            OUT.get().print("Submit failed. Failed to get details for the given extension");
+            throw new FalconCLIException("Submit failed. Failed to get details for the given extension");
+        }
+        String extensionType;
         String extensionBuildLocation;
         try {
-            JSONObject extensionDetailsJson = new JSONObject(responseJson.get("detail").toString());
-            extensionType = ExtensionType.valueOf(extensionDetailsJson.get("type").toString().toUpperCase());
-            extensionBuildLocation = extensionDetailsJson.get("location").toString();
+            extensionType = responseJson.get("type").toString();
+            extensionBuildLocation = responseJson.get("location").toString();
         } catch (JSONException e) {
             OUT.get().print("Error. " + extensionName + " not found ");
-            return false;
+            throw new FalconCLIException("Submit failed. Failed to get details for the given extension");
         }
 
-        List<Entity> entities;
-        if (extensionType.equals(ExtensionType.CUSTOM)) {
+        List<Entity> entities = null;
+        if (!extensionType.equals(ExtensionType.CUSTOM.name())) {
             try {
                 entities = ExtensionHandler.loadAndPrepare(extensionName, jobName, configStream,
                         extensionBuildLocation);
             } catch (Exception e) {
                 OUT.get().println("Error in building the extension");
-                return false;
+                throw new FalconCLIException("Submit failed. Failed to get details for the given extension");
             }
             if (entities == null || entities.isEmpty()) {
                 OUT.get().println("No entities got built");
-                return false;
+                throw new FalconCLIException("Submit failed. Failed to get details for the given extension");
             }
             try {
```

Comment:
Apologies as I just noticed this. It would be better if this is part of the ExtensionHandler itself. Client should be as light as possible.